// allure-store/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  AllureStore: () => AllureStore,
  CompiledAllureReader: () => CompiledAllureReader,
  FileAllureReader: () => FileAllureReader,
  FileAllureWriter: () => FileAllureWriter,
  FileReportSource: () => FileReportSource,
  HtmlReportSource: () => HtmlReportSource,
  UrlReportSource: () => UrlReportSource,
  fromConfig: () => fromConfig,
  fromDirectory: () => fromDirectory,
  fromReport: () => fromReport
});
module.exports = __toCommonJS(index_exports);

// src/AllureStore.ts
var import_promises4 = __toESM(require("node:fs/promises"));

// src/errors.ts
function resolveOnError(onError) {
  if (onError === void 0 || onError === "throw") {
    return (e) => {
      throw e;
    };
  } else if (onError === "ignore") {
    return () => {
    };
  } else {
    return onError;
  }
}
var AllureStoreError = class extends Error {
  cause;
  constructor(message, cause) {
    super(message, { cause });
    this.name = "AllureStoreError";
  }
};

// src/CompiledAllureReader.ts
var CompiledAllureReader = class {
  #source;
  #onError;
  constructor(config) {
    this.#source = config.source;
    this.#onError = resolveOnError(config.onError ?? "throw");
  }
  async init() {
    var _a, _b;
    await ((_b = (_a = this.#source).init) == null ? void 0 : _b.call(_a));
  }
  async getContainerIds() {
    return [];
  }
  async getResultIds() {
    const files = await this.#source.listFiles("data/test-cases");
    return files.map((f) => {
      const basename = f.split("/").pop();
      return basename.replace(/\.json$/, "");
    });
  }
  async readResult(id) {
    const content = await this.#source.readFile(`data/test-cases/${id}.json`);
    if (!content) return null;
    try {
      return mapCompiledTestCase(JSON.parse(content));
    } catch (error) {
      this.#handleError(`Failed to parse compiled test case: ${id}`, error);
      return null;
    }
  }
  async readContainer(_id) {
    return null;
  }
  async readCategories() {
    return null;
  }
  async readEnvironmentInfo() {
    const content = await this.#source.readFile("widgets/environment.json");
    if (!content) return null;
    try {
      const entries = JSON.parse(content);
      const result = {};
      for (const entry of entries) {
        result[entry.name] = entry.values.join(",");
      }
      return result;
    } catch (error) {
      this.#handleError("Failed to parse environment info", error);
      return null;
    }
  }
  async readExecutorInfo() {
    const content = await this.#source.readFile("widgets/executors.json");
    if (!content) return null;
    try {
      const executors = JSON.parse(content);
      return executors[0] ?? null;
    } catch (error) {
      this.#handleError("Failed to parse executor info", error);
      return null;
    }
  }
  #handleError(message, cause) {
    this.#onError(new AllureStoreError(message, cause));
  }
};
function mapCompiledTestCase(source) {
  const labels = [...source.labels ?? []];
  addSyntheticLabels(labels, source);
  return {
    uuid: source.uid,
    historyId: source.historyId,
    name: source.name,
    fullName: source.fullName,
    start: source.time.start,
    stop: source.time.stop,
    ...source.descriptionHtml ? { descriptionHtml: source.descriptionHtml } : {},
    stage: "finished",
    status: source.status,
    ...mapStatusDetails(source),
    labels,
    ...source.links ? { links: source.links } : {},
    ...source.parameters ? { parameters: source.parameters } : {},
    steps: assembleSteps(source)
  };
}
function mapStatusDetails(source) {
  if (!source.statusMessage && !source.statusTrace) return {};
  return {
    statusDetails: {
      ...source.statusMessage ? { message: source.statusMessage } : {},
      ...source.statusTrace ? { trace: source.statusTrace } : {}
    }
  };
}
var SYNTHETIC_BOOLEAN_LABELS = [
  ["flaky", "allure2.flaky"],
  ["retry", "allure2.retry"],
  ["hidden", "allure2.hidden"],
  ["retriesStatusChange", "allure2.retriesStatusChange"],
  ["newFailed", "allure2.newFailed"],
  ["newBroken", "allure2.newBroken"],
  ["newPassed", "allure2.newPassed"]
];
function addSyntheticLabels(labels, source) {
  for (const [key, labelName] of SYNTHETIC_BOOLEAN_LABELS) {
    if (source[key]) {
      labels.push({ name: labelName, value: "true" });
    }
  }
  if (source.retriesCount && source.retriesCount > 0) {
    labels.push({ name: "allure2.retriesCount", value: String(source.retriesCount) });
  }
}
function assembleSteps(source) {
  var _a, _b;
  const before = (source.beforeStages ?? []).map(mapStep);
  const test = ((_b = (_a = source.testStage) == null ? void 0 : _a.steps) == null ? void 0 : _b.map(mapStep)) ?? [];
  const after = (source.afterStages ?? []).map(mapStep);
  return [...before, ...test, ...after];
}
function mapStep(step) {
  var _a, _b, _c, _d, _e;
  return {
    name: step.name,
    start: ((_a = step.time) == null ? void 0 : _a.start) ?? 0,
    stop: ((_b = step.time) == null ? void 0 : _b.stop) ?? 0,
    stage: "finished",
    status: step.status,
    ...mapStatusDetails(step),
    ...((_c = step.steps) == null ? void 0 : _c.length) ? { steps: step.steps.map(mapStep) } : {},
    ...((_d = step.attachments) == null ? void 0 : _d.length) ? { attachments: step.attachments.map(mapAttachment) } : {},
    ...((_e = step.parameters) == null ? void 0 : _e.length) ? { parameters: step.parameters } : {}
  };
}
function mapAttachment(att) {
  return {
    name: att.name,
    type: att.type,
    source: att.source,
    ...att.size == null ? {} : { size: att.size }
  };
}

// src/CompiledReportSource.ts
var import_promises = __toESM(require("node:fs/promises"));
var import_node_path = __toESM(require("node:path"));
function stripFileProtocol(input) {
  return input.startsWith("file://") ? input.slice(7) : input;
}
var FileReportSource = class {
  #dir;
  constructor(dir) {
    this.#dir = dir;
  }
  async readFile(relativePath) {
    const resolved = import_node_path.default.resolve(this.#dir, relativePath);
    if (!resolved.startsWith(import_node_path.default.resolve(this.#dir) + import_node_path.default.sep)) return null;
    try {
      return await import_promises.default.readFile(resolved, "utf8");
    } catch {
      return null;
    }
  }
  async listFiles(prefix) {
    const dir = import_node_path.default.resolve(this.#dir, prefix);
    if (!dir.startsWith(import_node_path.default.resolve(this.#dir) + import_node_path.default.sep)) return [];
    try {
      const entries = await import_promises.default.readdir(dir);
      return entries.map((e) => `${prefix}/${e}`);
    } catch {
      return [];
    }
  }
};
var UrlReportSource = class {
  #baseUrl;
  constructor(baseUrl) {
    this.#baseUrl = baseUrl.replace(/\/+$/, "");
  }
  async readFile(relativePath) {
    try {
      const response = await fetch(`${this.#baseUrl}/${relativePath}`);
      if (!response.ok) return null;
      return await response.text();
    } catch {
      return null;
    }
  }
  // Only 'data/test-cases' is supported — URL-based reports have no directory listing,
  // so test case UIDs are discovered via 'widgets/status-chart.json' instead.
  async listFiles(prefix) {
    if (prefix === "data/test-cases") {
      return this.#listTestCaseFiles();
    }
    return [];
  }
  async #listTestCaseFiles() {
    const content = await this.readFile("widgets/status-chart.json");
    if (!content) return [];
    try {
      const entries = JSON.parse(content);
      return entries.map((e) => `data/test-cases/${e.uid}.json`);
    } catch {
      return [];
    }
  }
};
var HTML_DATA_PATTERN = /\bd\(["']([^"']+)["'],\s*["']([^"']+)["']\)/;
var HTML_DATA_PATTERN_GLOBAL = new RegExp(HTML_DATA_PATTERN, "g");
var HtmlReportSource = class {
  #input;
  #prefetched;
  #files = /* @__PURE__ */ new Map();
  constructor(fileOrUrl, prefetchedHtml) {
    this.#input = fileOrUrl;
    this.#prefetched = prefetchedHtml;
  }
  async init() {
    const html = this.#prefetched ?? await this.#fetchHtml();
    if (!html) return;
    for (const match of html.matchAll(HTML_DATA_PATTERN_GLOBAL)) {
      this.#files.set(match[1], Buffer.from(match[2], "base64").toString("utf8"));
    }
  }
  async readFile(relativePath) {
    return this.#files.get(relativePath) ?? null;
  }
  async listFiles(prefix) {
    return [...this.#files.keys()].filter((k) => k.startsWith(prefix));
  }
  async #fetchHtml() {
    const input = this.#input;
    if (input.startsWith("http://") || input.startsWith("https://")) {
      try {
        const response = await fetch(input);
        if (!response.ok) return null;
        return await response.text();
      } catch {
        return null;
      }
    }
    try {
      return await import_promises.default.readFile(stripFileProtocol(input), "utf8");
    } catch {
      return null;
    }
  }
};
function containsHtmlDataCalls(content) {
  return HTML_DATA_PATTERN.test(content);
}

// src/FileAllureReader.ts
var import_promises2 = __toESM(require("node:fs/promises"));
var import_node_path2 = __toESM(require("node:path"));
var properties = __toESM(require("properties"));
var FileAllureReader = class {
  #resultsDirectory;
  #onError;
  #scanIdsPromise;
  #containerIds;
  #resultIds;
  constructor(config) {
    this.#resultsDirectory = config.resultsDirectory;
    this.#onError = resolveOnError(config.onError || "throw");
  }
  async getContainerIds() {
    if (!this.#scanIdsPromise) {
      this.#scanIdsPromise = this.#scanIds();
    }
    await this.#scanIdsPromise;
    return this.#containerIds || [];
  }
  async getResultIds() {
    if (!this.#scanIdsPromise) {
      this.#scanIdsPromise = this.#scanIds();
    }
    await this.#scanIdsPromise;
    return this.#resultIds || [];
  }
  async readResult(id) {
    const filePath = import_node_path2.default.join(this.#resultsDirectory, `${id}-result.json`);
    return this.#parseJSON(filePath);
  }
  async readContainer(id) {
    const filePath = import_node_path2.default.join(this.#resultsDirectory, `${id}-container.json`);
    return this.#parseJSON(filePath);
  }
  async readCategories() {
    const filePath = import_node_path2.default.join(this.#resultsDirectory, "categories.json");
    return this.#parseJSON(filePath);
  }
  async readEnvironmentInfo() {
    const filePath = import_node_path2.default.join(this.#resultsDirectory, "environment.properties");
    const content = await this.#readFileIfExists(filePath);
    if (!content) return null;
    return properties.parse(content, { sections: false, comments: "#", separators: "=", unicode: true });
  }
  async readExecutorInfo() {
    const filePath = import_node_path2.default.join(this.#resultsDirectory, "executor.json");
    return this.#parseJSON(filePath);
  }
  /**
   * Helper method to read and parse JSON files.
   */
  async #parseJSON(filePath) {
    try {
      const content = await this.#readFileIfExists(filePath);
      return content ? JSON.parse(content) : null;
    } catch (error) {
      this.#handleError(`Failed to parse JSON file: ${filePath}`, error);
      return null;
    }
  }
  /**
   * Helper method to safely read a file, returning `null` if it doesn't exist.
   */
  async #readFileIfExists(filePath) {
    try {
      return await import_promises2.default.readFile(filePath, "utf8");
    } catch (error) {
      this.#handleError(`Failed to read file: ${filePath}`, error);
      return null;
    }
  }
  /**
   * Helper method to list the contents of the results directory and cache the IDs.
   */
  async #scanIds() {
    let files = [];
    try {
      files = await import_promises2.default.readdir(this.#resultsDirectory);
    } catch (error) {
      this.#handleError(`Failed to list directory: ${this.#resultsDirectory}`, error);
      files = [];
    }
    this.#containerIds = files.filter((f) => f.endsWith("-container.json")).map((f) => import_node_path2.default.basename(f, "-container.json"));
    this.#resultIds = files.filter((f) => f.endsWith("-result.json")).map((f) => import_node_path2.default.basename(f, "-result.json"));
  }
  #handleError(message, cause) {
    this.#onError(new AllureStoreError(message, cause));
  }
};

// src/FileAllureWriter.ts
var import_promises3 = __toESM(require("node:fs/promises"));
var import_node_path3 = __toESM(require("node:path"));
var properties2 = __toESM(require("properties"));
var FileAllureWriter = class {
  #config;
  #onError;
  constructor(config) {
    this.#config = config;
    this.#onError = resolveOnError(config.onError || "throw");
  }
  async init() {
    const { resultsDirectory, overwrite } = this.#config;
    try {
      const directoryExists = await import_promises3.default.access(resultsDirectory).then(
        () => true,
        // throw if not ENOENT
        (error) => {
          if ((error == null ? void 0 : error.code) !== "ENOENT") throw error;
          return false;
        }
      );
      if (overwrite && directoryExists) {
        await import_promises3.default.rm(resultsDirectory, { recursive: true });
      }
      await import_promises3.default.mkdir(resultsDirectory, { recursive: true });
    } catch (error) {
      this.#handleError(`Failed to initialize results directory: ${resultsDirectory}`, error);
    }
  }
  async writeCategories(categories) {
    const filePath = this.#buildPath("categories.json");
    try {
      await writeJson(filePath, categories, regexpAwareStringifier);
    } catch (error) {
      this.#handleError(`Failed to write categories: ${filePath}`, error);
    }
  }
  async writeContainer(container) {
    const filePath = this.#buildPath(`${container.uuid}-container.json`);
    try {
      await writeJson(filePath, container);
    } catch (error) {
      this.#handleError(`Failed to write container: ${filePath}`, error);
    }
  }
  async writeEnvironmentInfo(info) {
    const filePath = this.#buildPath("environment.properties");
    try {
      const text = properties2.stringify(info, { unicode: true });
      await import_promises3.default.writeFile(filePath, text + "\n");
    } catch (error) {
      this.#handleError(`Failed to write environment info: ${filePath}`, error);
    }
  }
  async writeExecutorInfo(info) {
    const filePath = this.#buildPath("executor.json");
    try {
      await writeJson(filePath, info);
    } catch (error) {
      this.#handleError(`Failed to write executor info: ${filePath}`, error);
    }
  }
  async writeResult(result) {
    const filePath = this.#buildPath(`${result.uuid}-result.json`);
    try {
      await writeJson(filePath, result);
    } catch (error) {
      this.#handleError(`Failed to write result: ${filePath}`, error);
    }
  }
  #buildPath(name) {
    return import_node_path3.default.join(this.#config.resultsDirectory, name);
  }
  #handleError(message, cause) {
    this.#onError(new AllureStoreError(message, cause));
  }
};
async function writeJson(filePath, data, stringifier) {
  const json = JSON.stringify(data, stringifier);
  await import_promises3.default.writeFile(filePath, json + "\n");
}
function regexpAwareStringifier(_key, value) {
  return value instanceof RegExp ? value.source : value;
}

// src/AllureStore.ts
var AllureStore = class _AllureStore {
  #reader;
  #writer;
  constructor(config) {
    this.#reader = config.reader;
    this.#writer = config.writer;
  }
  static async fromConfig(config) {
    var _a, _b, _c, _d;
    await Promise.all([(_b = (_a = config.reader).init) == null ? void 0 : _b.call(_a), (_d = (_c = config.writer) == null ? void 0 : _c.init) == null ? void 0 : _d.call(_c)]);
    return new _AllureStore(config);
  }
  static async fromDirectory(resultsDirectory, config = {}) {
    const reader = new FileAllureReader({ resultsDirectory, onError: config.onError });
    const writer = new FileAllureWriter({ resultsDirectory, onError: config.onError, overwrite: config.overwrite });
    return _AllureStore.fromConfig({ reader, writer });
  }
  static async fromReport(input, options = {}) {
    const localPath = stripFileProtocol(input);
    if (!localPath.startsWith("http://") && !localPath.startsWith("https://")) {
      const stat = await import_promises4.default.stat(localPath).catch(() => null);
      if (stat == null ? void 0 : stat.isDirectory()) {
        return _AllureStore.fromConfig({ reader: new CompiledAllureReader({ source: new FileReportSource(localPath), ...options }) });
      }
      if (stat == null ? void 0 : stat.isFile()) {
        return _AllureStore.fromConfig({ reader: new CompiledAllureReader({ source: new HtmlReportSource(localPath), ...options }) });
      }
      throw new Error(`Cannot resolve report input: ${input}`);
    }
    const response = await fetch(localPath);
    if (!response.ok) {
      throw new Error(`Failed to fetch report: ${response.status} ${response.statusText}`);
    }
    const html = await response.text();
    if (containsHtmlDataCalls(html)) {
      return _AllureStore.fromConfig({ reader: new CompiledAllureReader({ source: new HtmlReportSource(localPath, html), ...options }) });
    }
    return _AllureStore.fromConfig({ reader: new CompiledAllureReader({ source: new UrlReportSource(localPath), ...options }) });
  }
  //#region Reading methods
  async getAllResults() {
    const [containerIds, resultIds] = await Promise.all([
      this.#reader.getContainerIds(),
      this.#reader.getResultIds()
    ]);
    const containerMap = await this.#buildContainerMap(containerIds);
    const rawResults = await Promise.all(resultIds.map((id) => this.#reader.readResult(id)));
    const validResults = rawResults.filter((r) => r != null);
    return validResults.map((result) => this.#mergeContainers(result, containerMap));
  }
  async getLatestResults() {
    const all = await this.getAllResults();
    const map = /* @__PURE__ */ new Map();
    for (const result of all) {
      const existing = map.get(result.historyId);
      if (!existing || result.stop > existing.stop) {
        map.set(result.historyId, result);
      }
    }
    return [...map.values()];
  }
  async getCategories() {
    return this.#reader.readCategories();
  }
  async getEnvironment() {
    return this.#reader.readEnvironmentInfo();
  }
  async getExecutor() {
    return this.#reader.readExecutorInfo();
  }
  async getContainer(id) {
    return this.#reader.readContainer(id);
  }
  async getResult(id) {
    return this.#reader.readResult(id);
  }
  //#endregion
  //#region Writing methods
  async writeCategories(categories) {
    return this.#requireWriter().writeCategories(categories);
  }
  async writeEnvironmentInfo(info) {
    return this.#requireWriter().writeEnvironmentInfo(info);
  }
  async writeExecutorInfo(info) {
    return this.#requireWriter().writeExecutorInfo(info);
  }
  async writeResult(result) {
    return this.#requireWriter().writeResult(result);
  }
  async writeContainer(container) {
    return this.#requireWriter().writeContainer(container);
  }
  #requireWriter() {
    if (!this.#writer) {
      throw new Error("This store is read-only (no writer configured)");
    }
    return this.#writer;
  }
  //#endregion
  //#region Private helper methods
  async #buildContainerMap(containerIds) {
    const containerMap = /* @__PURE__ */ new Map();
    await Promise.all(
      containerIds.map(async (cid) => {
        const container = await this.#reader.readContainer(cid);
        if (container) {
          containerMap.set(container.uuid, container);
        }
      })
    );
    return containerMap;
  }
  #mergeContainers(result, containerMap) {
    const ancestors = this.#findAncestorContainers(result.uuid, containerMap);
    const merged = { ...result, steps: this.#mergeSteps(result, ancestors) };
    return merged;
  }
  #findAncestorContainers(startId, containerMap) {
    const parentFor = /* @__PURE__ */ new Map();
    for (const container of containerMap.values()) {
      for (const child of container.children) {
        parentFor.set(child, container.uuid);
      }
    }
    const ancestors = [];
    let currentId = startId;
    while (currentId) {
      const parentId = parentFor.get(currentId);
      if (!parentId) break;
      const parentContainer = containerMap.get(parentId);
      if (!parentContainer) break;
      ancestors.push(parentContainer);
      currentId = parentId;
    }
    return ancestors.reverse();
  }
  #mergeSteps(result, ancestors) {
    const beforeSteps = ancestors.flatMap((a) => a.befores ?? []);
    const afterSteps = [...ancestors].reverse().flatMap((a) => a.afters ?? []);
    return [...beforeSteps, ...result.steps ?? [], ...afterSteps];
  }
  //#endregion
};

// src/index.ts
async function fromConfig(options) {
  return AllureStore.fromConfig(options);
}
async function fromDirectory(resultsDirectory, options) {
  return AllureStore.fromDirectory(resultsDirectory, options);
}
async function fromReport(input, options) {
  return AllureStore.fromReport(input, options);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AllureStore,
  CompiledAllureReader,
  FileAllureReader,
  FileAllureWriter,
  FileReportSource,
  HtmlReportSource,
  UrlReportSource,
  fromConfig,
  fromDirectory,
  fromReport
});
